app/services/credits/applied_prepaid_credits_service.rb

```diff
@@ -0,0 +1,151 @@
+      @invoice = invoice
+      @wallets = wallets
+
+      super(nil)
+    end
+
+    def call
+      result.prepaid_credit_amount_cents ||= 0
+      result.wallet_transactions ||= []
+
+      if wallets_already_applied?
+        return result.service_failure!(code: "already_applied", message: "Prepaid credits already applied")
+      end
+
+      # build a hash of remaining_amounts for each (fee_type and billable_metric_id) pair
+      remaining_amounts = calculate_amounts_for_fees_by_type_and_bm
+
+      ApplicationRecord.transaction do
+        wallets.each do |wallet|
+          break if remaining_amounts.values.all? { |v| BigDecimal(v.to_s) <= 0 }
+
+          # returns applied amount on the fees_by_type_and_bm
+          amount = applicable_wallet_amount(wallet, remaining_amounts)
+          next if amount <= 0
+
+          wallet_transaction = create_and_decrease!(wallet: wallet, amount_cents: amount, invoice:)
+          result.wallet_transactions << wallet_transaction
+
+          Utils::ActivityLog.produce(wallet_transaction, "wallet_transaction.created", after_commit: true)
+          after_commit { SendWebhookJob.perform_later("wallet_transaction.created", wallet_transaction) }
+
+          invoice.prepaid_credit_amount_cents += amount
+          result.prepaid_credit_amount_cents += amount
+        end
+
+        invoice.save! if invoice.changed?
+      end
+      result
+    end
+
+    private
+
+    attr_accessor :invoice, :wallets
+
+    def wallets_already_applied?
+      return false unless invoice
+      wallet_ids = wallets.map { |w| w.id }
+      invoice.wallet_transactions.exists?(wallet_id: wallet_ids)
+    end
+
+    # build a hash of [fee_type, billable_metric_id] => remaining_amount
+    def calculate_amounts_for_fees_by_type_and_bm
+      remaining = Hash.new(0)
+
+      invoice.fees.each do |f|
+        cap = f.sub_total_excluding_taxes_amount_cents +
+          f.taxes_precise_amount_cents -
+          f.precise_credit_notes_amount_cents
+
+        key = [f.fee_type, f.charge&.billable_metric_id]
+        remaining[key] += cap
+      end
+
+      remaining
+    end
+
+    def applicable_wallet_amount(wallet, remaining_amounts_by_fee_type_and_bm)
+      amount_left = wallet.balance_cents
+      applied = 0
+
+      eligible_fee_keys_for_wallet(wallet).each do |key|
+        break if amount_left <= 0
+        next if remaining_amounts_by_fee_type_and_bm[key].to_i <= 0
+
+        take = [remaining_amounts_by_fee_type_and_bm[key], amount_left].min
+        remaining_amounts_by_fee_type_and_bm[key] -= take
+        amount_left -= take
+        applied += take
+      end
+
+      applied
+    end
+
+    def eligible_fee_keys_for_wallet(wallet)
+      all_fees = invoice.fees
+
+      chosen = if wallet.limited_to_billable_metrics?
+        fees_for_wallet_limited_to_bm(wallet, all_fees)
+      elsif wallet.limited_fee_types?
+        all_fees.select { |f| wallet.allowed_fee_types.include?(f.fee_type) }
```

Comment:
is it possible to have this instead?`all_fees.where(type: wallet.allowed_fee_types)`